### ISSUE TITLE
ListView unable to observe its bindingContext properly

### ISSUE THREAD

**Issue by NordlingArt** [NativeScript/NativeScript#2793]
**Platform:** iOS
**TNS:** 2.3.0
**Git repo:** https://github.com/NordlingArt/tns-issue-listview

If I make a checklist to a `ListView`, I need to observe an object property called `selected`, which is set to `false` as default, so I can either show or hide the checkbox.

In the Git repo, I have set up a `ListView` and a `Repeater`. Both use the same items `observableArray` and they both interact with each other. However, only `Repeater` will update its UI properly when selecting other options than the initially selected one.

It seems like the `ListView` can only observe the value if the condition is `true` from the beginning. See the `selected` property below:

``` js
var options = new ObservableArray(
  new Observable({ index: 0, text: "Option 1", selected: true }),
  new Observable({ index: 1, text: "Option 2", selected: false }),
  new Observable({ index: 2, text: "Option 3", selected: false }),
  new Observable({ index: 3, text: "Option 4", selected: false })
);
```

Condition from the XML:

``` xml
<StackLayout col="1" class="checkbox" visibility="{{ selected ? 'visible' : 'collapsed' }}" />
```

I strongly suggest you take a look at the Git repo for a full demo of the issue.

**Comment by markosko** [comment]
I would try this change, using binding before expression
```xml
<StackLayout col="1" class="checkbox" visibility="{{selected, selected ? 'visible' : 'collapsed' }}" />
```

**Comment by NickIliev** [comment]
Hey @markosko ,

I have reproduced your issue and can confirm that your code is nested binding the **_selected_** property won't bind with ListView (but the same technique will work with Repeater). We will investigate further and come back with a resolution.

Code to reproduce this behavior (look in the client app link or use the following below):
```
  <GridLayout rows="auto, *, auto, auto, *">
    <Label row="0" text="ListView" class="title" />

    <!-- this WON'T work! -->
    <ListView row="1" items="{{ options }}" id="listView">
      <ListView.itemTemplate>
        <GridLayout columns="*, auto" id="{{ index, 'option-' + index }}" class="item" tap="selectOption">
          <Label col="0" text="{{ text }}" />
          <StackLayout col="1" class="checkbox" visibility="{{ selected ? 'visible' : 'collapsed' }}" />
        </GridLayout>
      </ListView.itemTemplate>
    </ListView>
    
    <StackLayout row="2" class="border" />
    
    <Label row="3" text="Repeater" class="title" />

    <!-- this will work! -->
    <Repeater row="4" items="{{ options }}" id="repeater">
      <Repeater.itemTemplate>
        <GridLayout columns="*, auto" id="{{ index, 'option-' + index }}" class="item" tap="selectOption">
          <Label col="0" text="{{ text }}" />
          <StackLayout col="1" class="checkbox" visibility="{{ selected ? 'visible' : 'collapsed' }}" />
        </GridLayout>
      </Repeater.itemTemplate>
    </Repeater>
    
  </GridLayout>
```
**_page.js_**


```
var Observable = require("data/observable").Observable;
var ObservableArray = require("data/observable-array").ObservableArray;

var options = new ObservableArray(
  new Observable({ index: 0, text: "Option 1", selected: true }),
  new Observable({ index: 1, text: "Option 2", selected: false }),
  new Observable({ index: 2, text: "Option 3", selected: false }),
  new Observable({ index: 3, text: "Option 4", selected: false })
);

var page, pageData = new Observable({
  options: options
});

exports.navigatingTo = function(args) {
  page = args.object;
  page.bindingContext = pageData;
};

exports.selectOption = function(args) {
  var index = args.object.id.replace("option-", "");
  
  options.forEach(function(val, idx, arr) {
    if(idx == index) val.selected = true;
      else val.selected = false;
  });
};
```

**Comment by NordlingArt** [comment]
@markosko - No difference in doing that unfortunately.

@NickIliev - Thanks for investigating!

**Comment by NickIliev** [comment]
Hey @NordlingArt 

At this very moment, this behaviour is by design for iOS list-view. So your workaround is the one applicable (to use repeater in situation when you need to control inner items visibility),

**Comment by NordlingArt** [comment]
@NickIliev - That's unfortunate, because a Repeater will freeze when loading items over a specific amount. Which is my case.

Also, if you unselect the already selected option, it works on that specific item. So why would it work if it's selected at first?

**Comment by NickIliev** [comment]
@NordlingArt 

I can confirm that indeed this is a bug as the exact same behaviour is noticed under Android as well! 
If the initial value of **_selected_** is set to false it's **NOT** respected in List-View but is working as expected in Repeater.
However, if the fix from @NordlingArt solution above is applied everything works in list-view.

Code to reproduce : [https://github.com/NickIliev/NativeScript-issues/tree/master/NativeScript/issue-2793/test](https://github.com/NickIliev/NativeScript-issues/tree/master/NativeScript/issue-2793/test)

**Comment by NordlingArt** [comment]
@NickIliev - Glad to have this bug confirmed! Patiently waiting for a fix 🙂

**Comment by NordlingArt** [comment]
@NickIliev - I have found the issue. It is pretty simple. In the XML:

**Will not work**
`visibility="{{ selected ? 'visible' : 'collapsed' }}"`

**But this will**
`opacity="{{ selected ? 1 : 0 }}"`

Such a simple fix, and it's been taking me months to realize 😳  Using this workaround, this might however create other issues if, for instance, I'm already using the opacity property for something else. Luckily this is not the case for me, but maybe to others.

**Comment by NathanaelA** [comment]
@NordlingArt - just so you are aware, opacity will make the item invisible; BUT you can still fully interact with it as if it was still visible...  Just something to consider in case this is something you can tap on.

**Comment by NordlingArt** [comment]
@NathanaelA - I believe if the opacity is set to 0, the interaction automatically gets disabled. At least on iOS. Cannot confirm it, but I've seen it happen before.

Though for me this is not an issue since I am toggling a subview. However, this seems like a bug that should be fixed rather soon.

**Comment by NathanaelA** [comment]
Interesting, I can confirm that on iOS if you set opacity = 0, it does appear to disable the control in 2.3.0
However in Android 2.4.0, opacity does not affect the controls ability to do anything...

**Comment by jacargentina** [comment]
@NickIliev I'm also facing the same iOS issue in latest (2.5.1) (Android works just fine). Any news on the fix ? Also, as a workaround, can i safely replace ListView -> Repeater ? Or there is some UI feature from ListView tha ti will miss ?

**Comment by jacargentina** [comment]
Debugging on the list-view-common.ts file on my use case, i've found the _onItemsChanged() is not called when the bound observable (the visibility conditional) changes.

So, the refresh() is also not called.

Anybody can give me a hand to fix this bug? Thanks!

**Comment by jacargentina** [comment]
Trying to get coded a test for this. Anybody can point me on how to run just 1 test?

Currently i'm doing `npm run test-ios` but this runs all of them, there is some way to just run the list view tests ?

**Comment by jacargentina** [comment]
I'm finding the main issue: the already rendered cells (on the iOS table for the ListView) **are not updated** when the observable for the visibility changes. May be it is time to call everyone involved on the file `list-view.ios.ts`

@hshristov @hamorphis @zh-m

**Comment by smarza** [comment]
@NordlingArt 
Hey guys!

Any fixes?

thanks

**Comment by nmongiya** [comment]
@smarza I am refreshing the listview after changing the variable value (I have a boolean variable that determines the visibility of the template) and this works for me on ios.

**Comment by NickIliev** [comment]
Issue resolved with tns-core-modules@4.1.0.

Test application can be found [here](https://github.com/NickIliev/NS-Issues-2018-II/tree/master/NativeScript/issue-2793)
